File: 2019/hs/src/Day1.hs
module Day1 (day1) where

import Control.Monad (guard)

fuel :: Int -> Int
fuel mass = mass `div` 3 - 2

fuelRec :: Int -> Int
fuelRec = go 0
  where
    go acc 0 = acc
    go acc new =
      let f = max 0 $ fuel new
      in go (acc + f) f

day1 :: String -> IO ()
day1 input = do
  let ls = lines input
  let part1 = sum $ map (fuel . read) ls
  let part2 = sum $ map (fuelRec . read) ls
  guard (part1 == 3455717 && part2 == 5180690)
  print part1
  print part2

File: 2019/rs/src/day1.rs
use std::cmp;
use std::fs::File;
use std::io::{self, BufRead};

pub fn day1(input_path: &str) {
    let file =
        File::open(input_path).unwrap_or_else(|e| panic!("Failed to open {}: {}", input_path, e));
    let weights = io::BufReader::new(file)
        .lines()
        .map(|l| {
            l.unwrap_or_else(|e| panic!("Failed to read line: {}", e))
                .trim()
                .parse::<i32>()
                .unwrap_or_else(|e| panic!("Failed to parse integer {}", e))
        })
        .collect::<Vec<i32>>();
    assert_eq!(
        weights.iter().copied().map(calc_fuel_simple).sum::<i32>(),
        3455717
    );
    assert_eq!(
        weights.iter().copied().map(calc_fuel_complex).sum::<i32>(),
        5180690
    );
}

fn calc_fuel_simple(weight: i32) -> i32 {
    cmp::max(weight / 3 - 2, 0)
}

fn calc_fuel_complex(mut weight: i32) -> i32 {
    let mut sum: i32 = 0;
    while weight > 0 {
        let fuel = calc_fuel_simple(weight);
        sum += fuel;
        weight = fuel;
    }
    sum
}
